core/primitives/src/time.rs

```diff
@@ -0,0 +1,146 @@
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::default::Default;
+use std::mem;
+use std::sync::{Arc, Mutex, Once};
+use std::time;
+
+use chrono;
+
+pub use chrono::Utc;
+pub use std::time::Instant;
+
+pub struct UtcProxy;
+pub struct InstantProxy;
+
+pub struct TimeTravelSingleton {
+    pub last_check_utc: chrono::DateTime<chrono::Utc>,
+    pub last_check_instant: time::Instant,
+    pub diff: i64,
+    pub rate: f64,
+    pub proxify: HashMap<FileLocation, bool>,
+}
+
+#[derive(Eq, PartialEq, Hash, Debug)]
+pub struct FileLocation {
+    pub file: Cow<'static, str>,
+    pub line: u32,
+}
+
+#[derive(Clone)]
+struct SingletonReader {
+    // Since we might be used in many threads, we need to protect
+    // concurrent access
+    inner: Arc<Mutex<TimeTravelSingleton>>,
+}
+
+impl Default for TimeTravelSingleton {
+    fn default() -> Self {
+        Self {
+            last_check_utc: chrono::Utc::now(),
+            last_check_instant: time::Instant::now(),
+            diff: 0,
+            rate: 1.0,
+            proxify: HashMap::new(),
+        }
+    }
+}
+
+impl TimeTravelSingleton {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn set(value: Self) {
+        SingletonReader::set(value)
+    }
+}
+
+static mut SINGLETON: *const SingletonReader = 0 as *const SingletonReader;
+static ONCE: Once = Once::new();
+
+impl SingletonReader {
+    fn get() -> SingletonReader {
+        unsafe {
+            ONCE.call_once(|| {
+                SingletonReader::set(TimeTravelSingleton::default());
+            });
+
+            (*SINGLETON).clone()
+        }
+    }
+
+    fn set(value: TimeTravelSingleton) {
+        let singleton = SingletonReader { inner: Arc::new(Mutex::new(value)) };
+        unsafe {
+            SINGLETON = mem::transmute(Box::new(singleton));
+        }
+    }
+}
```

Comment:
This code can be called from the closure within `get`, so we have two use cases
1. this is the first time the assignment happens, and we must replace the static singleton as there is no inner value yet
2. this is not our first store
I can see how we could improve on the code so that it does not leak memory in the second case. I removed this code in favor of `arc-swap`, though, so the concerns no longer apply.

for the record, the code here is taken from: https://stackoverflow.com/questions/27791532/how-do-i-create-a-global-mutable-singleton